drivers/opt3001/opt3001.c

```diff
@@ -0,0 +1,179 @@
+        return -OPT3001_ERROR_BUS;
+    }
+
+    reg = htons(reg);
+    if (reg != OPT3001_DID_VALUE) {
+        return -OPT3001_ERROR_DEV;
+    }
+
+    /* Set range number, mode of conversion and conversion time */
+    reg = OPT3001_CONFIG_RN_FSR;
+    reg |= OPT3001_REGS_CONFIG_MOC(OPT3001_CONFIG_M_SHUTDOWN);
+    reg |= OPT3001_REGS_CONFIG_CT(OPT3001_CONVERSION_TIME);
+
+    /* Configure for latched window-style comparison operation */
+    reg |= OPT3001_REGS_CONFIG_L;
+    reg &= ~OPT3001_REGS_CONFIG_POL;
+    reg &= ~OPT3001_REGS_CONFIG_ME;
+    reg &= ~OPT3001_REGS_CONFIG_FC_MASK;
+
+    reg = htons(reg);
+
+    if (i2c_write_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_init: Error setting device configuration\n");
+        return -OPT3001_ERROR_BUS;
+    }
+    i2c_release(DEV_I2C);
+
+    return OPT3001_OK;
+}
+
+int opt3001_reset(const opt3001_t *dev)
+{
+    uint16_t reg = OPT3001_CONFIG_RESET;
+    reg |= OPT3001_REGS_CONFIG_CT(OPT3001_CONVERSION_TIME);
+    reg = htons(reg);
+
+    /* Acquire exclusive access to the bus. */
+    i2c_acquire(DEV_I2C);
+    if (i2c_write_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_reset: Error setting device configuration\n");
+        return -OPT3001_ERROR_BUS;
+    }
+    i2c_release(DEV_I2C);
+    return OPT3001_OK;
+}
+
+int opt3001_set_active(const opt3001_t *dev)
+{
+    uint16_t reg;
+
+    i2c_acquire(DEV_I2C);
+
+    if (i2c_read_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_set_active: Error reading BUS!\n");
+        return -OPT3001_ERROR_BUS;
+    }
+
+    reg = htons(reg);
+    reg |= OPT3001_REGS_CONFIG_MOC(OPT3001_CONFIG_M_CONTINUOUS);
+
+    reg = htons(reg);
+    if (i2c_write_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+        i2c_release(DEV_I2C);
+        LOG_ERROR("opt3001_set_active: Error setting device configuration\n");
+        return -OPT3001_ERROR_BUS;
+    }
+
+    i2c_release(DEV_I2C);
+    return OPT3001_OK;
+}
+
+int opt3001_read_lux(const opt3001_t *dev, int32_t *convl)
+{
+    uint8_t exponent;
+    uint16_t reg, mantissa;
+    uint32_t conversion_time = 0;
+
+    i2c_acquire(DEV_I2C);
+
+    while (1) {
+        if (conversion_time >= OPT3001_CONVERSION_TIME_COMBINED) {
+            i2c_release(DEV_I2C);
+            LOG_ERROR("opt3001_read: Conversion in progress!\n");
+            return -OPT3001_ERROR;
+        }
+
+        if (i2c_read_regs(DEV_I2C, DEV_ADDR, OPT3001_REGS_CONFIG, &reg, 2, 0) < 0) {
+            i2c_release(DEV_I2C);
+            LOG_ERROR("opt3001_read: Error reading BUS!\n");
+            return -OPT3001_ERROR_BUS;
+        }
+
+        if (htons(reg) & OPT3001_REGS_CONFIG_CRF) {
+            break;
+        }
+
+        xtimer_usleep(1000);
```

Comment:
You should do `i2c_release(DEV_I2C);` before sleeping and then do `i2c_acquire(DEV_I2C);
` again afterwards.
This way you do not block the bus if some other threads wants to access a different I2C peripheral.

Also, if this will ever be implemented, it would allow the MCU to enter a deeper sleep state because the I2C peripheral is not in use. Unfortunately this is not the case yet.